pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,321 @@
+		}
+	}
+	return &secretObj
+}
+
+// writeEncryptionConfigurationSecretToSeed writes the secret which contains the EncryptionConfiguration
+// to the shoot namespace in the seed cluster
+func (b *Botanist) writeEncryptionConfigurationSecretToSeed(ecYamlBytes []byte, forcePlaintextSecrets bool) error {
+	secretObj := b.createEncryptionConfigurationSecret(common.EtcdEncryptionSecretName, b.Shoot.SeedNamespace, ecYamlBytes, forcePlaintextSecrets)
+	if _, err := b.K8sSeedClient.CreateSecretObject(secretObj, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+// writeEncryptionConfigurationSecretToGarden writes the secret which contains the EncryptionConfiguration
+// to the garden cluster
+func (b *Botanist) writeEncryptionConfigurationSecretToGarden(ecYamlBytes []byte) error {
+	secretNameInGardenCluster := b.calculateEtcdEncryptionSecretNameInGardenCluster()
+	secretObj := b.createEncryptionConfigurationSecret(secretNameInGardenCluster, b.Shoot.Info.Namespace, ecYamlBytes, false)
+	secretObj.ObjectMeta.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(b.Shoot.Info, gardenv1beta1.SchemeGroupVersion.WithKind("Shoot")),
+	}
+	if _, err := b.K8sGardenClient.CreateSecretObject(secretObj, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+// isEncryptionConfigurationConsistent checks whether the configuration is consistent in seed and garden
+func (b *Botanist) isEncryptionConfigurationConsistent(ecSeed *apiserverconfigv1.EncryptionConfiguration, ecGarden *apiserverconfigv1.EncryptionConfiguration) (bool, error) {
+	equal, err := encryptionconfiguration.Equals(ecSeed, ecGarden)
+	if err != nil {
+		return false, fmt.Errorf("error when comparing EncryptionConfiguration in seed cluster and garden cluster: %v", err)
+	}
+	if !equal {
+		return false, nil
+	}
+	consistent, err := encryptionconfiguration.IsConsistent(ecSeed)
+	if err != nil {
+		return false, fmt.Errorf("EncryptionConfiguration in seed cluster is not consistent: %v", err)
+	}
+	if !consistent {
+		return false, nil
+	}
+	return true, nil
+}
+
+// needToRewriteShootSecrets checks whether the secrets in the shoot need to
+// be rewritten, e.g. after a change to the EncryptionConfiguration
+func (b *Botanist) needToRewriteShootSecrets() (needToRewriteSecrets bool, err error) {
+	ecs, err := b.K8sSeedClient.GetSecret(b.Operation.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, fmt.Errorf("no EncryptionConfiguration found in seed cluster (%v)", b.Operation.Shoot.SeedNamespace)
+		}
+		return false, err
+	}
+	if ecs.Annotations == nil {
+		// no annotations found, rewrite secrets
+		return true, nil
+	}
+	lastSecretRewriteChecksum, ok := ecs.Annotations[common.EtcdEncryptionChecksumAnnotationName]
+	if !ok {
+		// no annotation found, rewrite secrets
+		return true, nil
+	}
+	if b.CheckSums[common.EtcdEncryptionSecretName] == lastSecretRewriteChecksum {
```

Comment:
Wouldn't it be better to have the check sum as a label and then get all secrets that have a label that doesn't match the checksum and only rewrite those? Would be way more efficient on network traffic + compute time.

Reply:
@adracus Do you suggest that, in every reconciliation loop, Gardener labels all existing secrets in the shoot cluster with the current etcd-encryption-secret checksum? Well, I actually like this idea, however, I don't think it would help much. The only reason why shoots secrets need to be rewritten is a change of the encryption secret itself. In this case, *all* secrets need to be rewritten anyways. I cannot imagine a case where only a subset of the existing secrets need to be considered. Do you?

Reply:
Ah, I see - yes, probably not that useful I guess.